### PR TITLE
(GH-1862) Specify preferred  algorithms for SSH transport connections

### PR DESCRIPTION
**Pull request by beechtom**
This adds the following SSH transport configuration options:

- `encryption-algorithms`
- `host-key-algorithms`
- `kex-algorithms`
- `mac-algorithms`

Each of these options accepts an array of preferred algorithms to use when
establishing a connection using the SSH transport. When a `*-algorithms`
option is set, `net-ssh` is configured to only attempt to establish a
connection to the target using algorithms in the list.

Each `*-algorithms` option accepts a special `defaults` item, which will
automatically add all of default, supported algorithms for `net-ssh` to
the list. For example, the following config:

```yaml
ssh:
  encryption-algorithms:
    - defaults
```

Would expand to the following:

```yaml
ssh:
  encryption-algorithms:
    - aes-256-ctr
    - aes-192-ctr
    - aes-128-ctr
```

Starting with the release of `net-ssh` 6.0, several algorithms were
deprecated and disabled by default. This prevented users from using
these algorithms when establishing connections with targets. These
options allow users to enable these deprecated algorithms.

!feature

* **Specify preferred algorithms for SSH transport connections**
  ([#1862](https://github.com/puppetlabs/bolt/issues/1862))

  Users can now specify a list of preferred algorithms to use when
  establishing connections with targets using the SSH transport with the
  `encryption-algorithms`, `host-key-algorithms`, `kex-algorithms`, and
  `mac-algorithms` config options. Each option accepts an array of
  algorithms and overrides the default list of preferred algorithms.
  You can read more about these options in the [Bolt configuration
  reference](https://puppet.com/docs/bolt/latest/bolt_configuration_reference.html#ssh).

### CODE REVIEW

**Comment by lucywyman** [comment]
I think these commits can be squashed.

File: lib/bolt/config/transport/ssh.rb [hunk 1]
```diff
@@ -12,94 +12,130 @@ class SSH < Base
...
+                                             "All supported, non-deprecated algorithms are enabled by default when "\
+                                             "this option is not used. To enable support for default algorithms "\
+                                             "when using this option, add 'defaults' to the list of supported "\
+                                             "algorithms." },
+          "host"                  => { type: String,
+                                       external: true,
+                                       desc: "Host name." },
+          "host-key-algorithms"   => { type: Array,
+                                       desc: "List of host key algorithms to support when establishing a connection "\
+                                             "with a target. Supported algorithms are defined by net-ssh "\
```
Comment (lucywyman) [inline lib/bolt/config/transport/ssh.rb:40]:
```suggestion
                                             "with a target. Supported algorithms are defined by the ruby net-ssh library "\
```

File: lib/bolt/transport/ssh/connection.rb [hunk 2]
```diff
@@ -246,6 +268,19 @@ def net_ssh_verifier(verifier)
...
+        # out any unsupported algorithms.
+        def net_ssh_algorithms(type, algorithms)
+          if algorithms.include?('defaults')
+            defaults = Net::SSH::Transport::Algorithms::DEFAULT_ALGORITHMS[type]
+            algorithms += defaults
+          end
+
+          known = Net::SSH::Transport::Algorithms::ALGORITHMS[type]
+
+          algorithms & known
```
Comment (lucywyman) [inline lib/bolt/transport/ssh/connection.rb:281]: Until #1861 is in and released, it'd be nice if we warned here that certain specified algorithms aren't supported at all.
```
supported, unsupported = algorithms.partition { |a| known.include?(a) }
@logger.warn("These #{type.to_s} algorithms aren't supported by the net-ssh library: #{unsupported}")
supported
```

Comment (beechtom) [reply lib/bolt/transport/ssh/connection.rb:281]: My only issue with warning about unsupported algorithms is that it will warn for every single target. If we want #1861 in as an alternate to warning, then I think we can just block this ticket until that's ready.

Comment (beechtom) [reply lib/bolt/transport/ssh/connection.rb:283]: Added the work for #1861 to this PR. Felt the error message should also list disabled algorithms and how to enable them, so it's closely tied to the work for #1862 .

**Comment by lucywyman** [comment]
It looks like we can add `Macs` to [this list](https://puppet.com/docs/bolt/latest/bolt_configuration_reference.html#openssh), but not KexAlgorithms

File: lib/bolt/transport/ssh/connection.rb [hunk 3]
```diff
@@ -275,6 +312,32 @@ def validate_ssh_version
             raise "Powershell over SSH requires OpenSSH server >= 7.9, target is running #{match[1]}"
           end
         end
+
+        def handle_net_ssh_error(err)
+          match = /could not settle on (?<algo>encryption|hmac|host_key|kex)/.match(err.message)
+
+          if match
+            enabled = Net::SSH::Transport::Algorithms::DEFAULT_ALGORITHMS[match[:algo].to_sym]
```
Comment (lucywyman) [inline lib/bolt/transport/ssh/connection.rb:320]: I think 'available' and 'deprecated' would be better words for these. They aren't what's configured by the user, and I think 'enabled' indicates something on the user-side.

**Comment by lucywyman** [comment]
I think all three commits can be squashed, this is all part of the same work.

File: schemas/bolt-transport-definitions.json [hunk 4]
```diff
@@ -49,29 +49,33 @@
...
+      "encryption-algorithms": { "$ref": "#/definitions/encryption-algorithms" },
+      "host":                  { "$ref": "#/definitions/host" },
+      "host-key-algorithms":   { "$ref": "#/definitions/host-key-algorithms" },
+      "host-key-check":        { "$ref": "#/definitions/host-key-check" },
+      "extensions":            { "$ref": "#/definitions/extensions" },
+      "interpreters":          { "$ref": "#/definitions/interpreters" },
+      "kex-algorithms":        { "$ref": "#/definitions/kex-algorithms" },
+      "load-config":           { "$ref": "#/definitions/load-config" },
+      "login-shell":           { "$ref": "#/definitions/login-shell" },
+      "mac-algorithms":        { "$ref": "#/definitions/mac-algorithms" },
```
Comment (lucywyman) [inline schemas/bolt-transport-definitions.json:65]:
```suggestion
      "hmac-algorithms":        { "$ref": "#/definitions/mac-algorithms" },
```

File: lib/bolt/config/transport/ssh.rb [hunk 5]
```diff
@@ -12,94 +12,132 @@ class SSH < Base
...
+          "connect-timeout"       => { type: Integer,
+                                       desc: "How long to wait when establishing connections." },
+          "copy-command"          => { external: true,
+                                       desc: "Command to use when copying files using ssh-command. "\
+                                             "Bolt runs `<copy-command> <src> <dest>`. **This option is "\
+                                             "experimental.**" },
+          "disconnect-timeout"    => { type: Integer,
+                                       desc: "How long to wait before force-closing a connection." },
+          "encryption-algorithms" => { type: Array,
+                                       desc: "List of encryption algorithms to support when establishing a "\
```
Comment (lucywyman) [inline lib/bolt/config/transport/ssh.rb:27]:
```suggestion
                                       desc: "List of encryption algorithms to use when establishing a "\
```

File: lib/bolt/config/transport/ssh.rb [hunk 6]
```diff
@@ -12,94 +12,132 @@ class SSH < Base
...
+                                             "connection with a target. Supported algorithms are "\
+                                             "defined by the Ruby net-ssh library and can be viewed "\
+                                             "[here](https://github.com/net-ssh/net-ssh#supported-algorithms). "\
+                                             "All supported, non-deprecated algorithms are enabled by default when "\
```
Comment (lucywyman) [inline lib/bolt/config/transport/ssh.rb:31]:
```suggestion
                                             "All supported, non-deprecated algorithms are available by default when "\
```

File: lib/bolt/config/transport/ssh.rb [hunk 7]
```diff
@@ -12,94 +12,132 @@ class SSH < Base
...
+                                             "this option is not used. To enable support for default algorithms "\
```
Comment (lucywyman) [inline lib/bolt/config/transport/ssh.rb:32]:
```suggestion
                                             "this option is not used. To use all default algorithms "\
```

Comment (lucywyman) [reply lib/bolt/config/transport/ssh.rb:32]: 'reference'? 'specify'? might also work. Paging @hestonhoffman

Comment (lucywyman) [reply lib/bolt/config/transport/ssh.rb:32]: Honestly overall this message is pretty good at communicating a weird situation. There's like...3 possible states for any algorithm to be in *in the net-ssh lib alone*, and then that's on top of whatever the user has configured locally.

File: lib/bolt/transport/ssh/connection.rb [hunk 8]
```diff
@@ -275,6 +312,32 @@ def validate_ssh_version
...
+            disabled = Net::SSH::Transport::Algorithms::ALGORITHMS[match[:algo].to_sym] - enabled
+
+            message = "#{err.message}\n"\
+                      "Supported algorithms are defined by the Ruby net-ssh library. All supported, "\
+                      "non-deprecated algorithms are enabled by default. Disabled algorithms can be "\
+                      "enabled by setting the '#{match[:algo].gsub('_', '-')}-algorithms' config option.\n"\
+                      "Enabled algorithms: #{enabled.join(',')}\n"\
```
Comment (nicklewis) [inline lib/bolt/transport/ssh/connection.rb:327]: In practice, it's hard to follow how this information relates to the error when it occurs. Maybe it makes sense in the case where you *haven't* specified a list of preferred algorithms, but when you did, it's weird that it shows the list that you *could* have selected.

For context, with `encryption-algorithms: [ "blowfish-ctr" ]`
```
  could not settle on encryption_client algorithm
  Server encryption_client preferences: chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com,aes128-cbc,aes192-cbc,aes256-cbc,blowfish-cbc,cast128-cbc,3des-cbc
  Client encryption_client preferences: blowfish-ctr
  Supported algorithms are defined by the Ruby net-ssh library. All supported, non-deprecated algorithms are enabled by default. Disabled algorithms can be enabled by setting the 'encryption-algorithms' config option.
  Enabled algorithms: aes256-ctr,aes192-ctr,aes128-ctr
  Disabled algorithms: aes256-cbc,aes192-cbc,aes128-cbc,rijndael-cbc@lysator.liu.se,blowfish-ctr,blowfish-cbc,cast128-ctr,cast128-cbc,3des-ctr,3des-cbc,idea-cbc,none
```
